### PR TITLE
SymbolBlock Serialization

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##

Serialization effort for MXNet and PyTorch Symbol blocks

### CODE REVIEW

File: mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/Symbol.java [hunk 1]
```diff
@@ -69,6 +69,18 @@ public static Symbol load(MxNDManager manager, String path) {
...
     }
 
+    /**
+     * Loads a symbol from a json string.
+     *
+     * @param manager the manager to load the symbol to
+     * @param json the json string of the symbol.
+     * @return the new symbol
+     */
+    public static Symbol loadString(MxNDManager manager, String json) {
```
Comment (stu1130) [inline mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/Symbol.java:79]: maybe call it loadSymbolJson or loadJson?

File: pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_inference.cc [hunk 2]
```diff
@@ -81,7 +94,7 @@ JNIEXPORT jlong JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_moduleLoad__Ljava
 }
 
 JNIEXPORT void JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_moduleWrite(
-    JNIEnv* env, jobject jthis, jlong module_handle, jobject jos, jbyteArray arr) {
+    JNIEnv* env, jobject jthis, jlong module_handle, jobject jos, jbyteArray arr, jboolean writeSize) {
```
Comment (stu1130) [inline pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_inference.cc:97]: jwrite_size

File: pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_inference.cc [hunk 3]
```diff
@@ -97,6 +110,16 @@ JNIEXPORT void JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_moduleWrite(
     env->ThrowNew(ENGINE_EXCEPTION_CLASS, "The write method in OutputStream is not found");
     return;
   }
+  if (writeSize) {
+      auto jbytes = env->NewByteArray(4);
+      auto length = str.size();
+      char bytes[4];
+      for (int i = 0; i < 4; i++) {
+          bytes[i] = (int)(length >> (24 - 8 * i) & 0XFF);
```
Comment (stu1130) [inline pytorch/pytorch-native/src/main/native/ai_djl_pytorch_jni_PyTorchLibrary_inference.cc:118]: static_cast<int>

File: mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxSymbolBlock.java [hunk 4]
```diff
@@ -253,6 +253,10 @@ public Shape getParameterShape(String name, Shape[] inputShapes) {
     @Override
     public void saveParameters(DataOutputStream os) throws IOException {
         os.writeByte(VERSION);
+        String json = symbol.toJsonString();
```
Comment (frankfliu) [inline mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxSymbolBlock.java:256]: Why not use os.writeUTF()?

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtSymbolBlock.java [hunk 5]
```diff
@@ -67,6 +71,19 @@ public PtSymbolBlock(PtNDManager manager, long handle) {
         first = true;
     }
 
+    /**
+     * Constructs an Empty {@code PtSymbolBlock}.
+     *
+     * @param manager the manager to use for the block
+     */
+    public PtSymbolBlock(PtNDManager manager) {
```
Comment (frankfliu) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtSymbolBlock.java:79]: This constructor is not been used, any plan to use this?

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java [hunk 6]
```diff
@@ -462,15 +462,15 @@ native void torchBackward(
...
+    native long moduleLoad(InputStream is, int[] device, byte[] buffer, long size);
 
     native void moduleEval(long handle);
 
     native void moduleTrain(long handle);
 
     native long moduleForward(long moduleHandle, long[] iValueHandles, boolean isTrain);
 
-    native void moduleWrite(long moduleHandle, OutputStream os, byte[] buffer);
+    native void moduleWrite(long moduleHandle, OutputStream os, byte[] buffer, boolean writeSize);
```
Comment (frankfliu) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java:473]: We always write with size, any case need false?

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java [hunk 7]
```diff
@@ -1378,22 +1380,34 @@ public static PtSymbolBlock loadModule(
...
-        long handle =
-                PyTorchLibrary.LIB.moduleLoad(
-                        is,
-                        new int[] {
-                            PtDeviceType.toDeviceType(device),
-                            device.equals(Device.cpu()) ? -1 : device.getDeviceId()
-                        },
-                        buf);
+    public static PtSymbolBlock loadModule(
+            PtNDManager manager, InputStream is, Device device, boolean hasSize)
```
Comment (frankfliu) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java:1384]: Why we need hasSize?

File: mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxSymbolBlock.java (hunk 4 above)
Comment (lanking520) [reply mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxSymbolBlock.java:256]: It's too long and write UTF only has 2 byte length

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtSymbolBlock.java (hunk 5 above)
Comment (lanking520) [reply pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtSymbolBlock.java:79]: Yes, in the blockfactory

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java (hunk 6 above)
Comment (lanking520) [reply pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java:473]: If anytime user are looking to save as PT file and use in Python

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java (hunk 7 above)
Comment (lanking520) [reply pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/JniUtils.java:1384]: Check if model has size in the front

**Review by frankfliu — APPROVED** [review]

**Pull request merged**